src/core/io/src/4C_io_input_spec_builders.hpp

```diff
@@ -126,6 +126,40 @@ namespace Core::IO
       return PrettyTypeName<T>{}();
     }
 
+    /**
+     * A small wrapper around a vector that provides a few convenience functions related to storing
+     * matched node IDs.
+     */
+    class Matches
```

Comment:
I implemented much better tracing of the match process with nice error messages. Since this PR contains enough non-trivial code already, I would push that in a follow-up. The whole `Matches` class will change to something much more powerful but for the initial version, it gets the job done.